File: CCButtonUse/CCButtonUse/ViewController.m
//
//	ViewController.m
//		CCButtonUse
//		Chen Chen @ Apirl 13th, 2015
//

#import "ViewController.h"

/// 按钮功能枚举数组
typedef NS_ENUM(NSInteger, kButtonType) {
    /// 上移按钮
    kButtonTypeTop = 10,
    /// 下移按钮
    kButtonTypeBottom,
    /// 左移按钮
    kButtonTypeLeft,
    /// 右移按钮
    kButtonTypeRight,
    /// 左转按钮
    kButtonTypeLeftRotate,
    /// 右转按钮
    kButtonTypeRightRotate,
    /// 放大按钮
    kButtonTypePlus,
    /// 缩小按钮
    kButtonTypeMinus,
};

/// 移动量
static const CGFloat kMovingDistance = 10.0;

@interface ViewController ()

/// 头像照片
@property (weak, nonatomic) IBOutlet UIButton *imageButton;

@end

@implementation ViewController

#pragma mark - 系统方法

/**
 *  视图已经加载方法
 */
- (void)viewDidLoad {
    
    [super viewDidLoad];
}

#pragma mark - 图片变化方法

/**
 *  移动图片方法
 */
- (IBAction)moveButtonDidClick:(UIButton *)button {
    
    [UIView beginAnimations:nil context:nil];
    [UIView setAnimationDuration:1.0];
    
    CGRect tempFrame = self.imageButton.frame;
    switch (button.tag) {
        case kButtonTypeTop:
            tempFrame.origin.y -= kMovingDistance;
            break;
            
        case kButtonTypeBottom:
            tempFrame.origin.y += kMovingDistance;
            break;
            
        case kButtonTypeLeft:
            tempFrame.origin.x -= kMovingDistance;
            break;
            
        case kButtonTypeRight:
            tempFrame.origin.x += kMovingDistance;
            break;
    }
    
    self.imageButton.frame = tempFrame;
    [UIView commitAnimations];
}

/**
 *  旋转图片方法
 */
- (IBAction)rotateButtonDidClick:(UIButton *)button {
    

    [UIView animateWithDuration:2.0 animations:^{
        switch (button.tag) {
            case kButtonTypeLeftRotate:
                self.imageButton.transform = CGAffineTransformRotate(self.imageButton.transform, -M_PI_4);
                break;
                
            case kButtonTypeRightRotate:
                self.imageButton.transform = CGAffineTransformRotate(self.imageButton.transform, M_PI_4);
                break;
        }
    }];
}

/**
 *  放缩图片方法
 */
- (IBAction)scalingButtonDidClick:(UIButton *)button {
    
    [UIView animateWithDuration:1.0 animations:^{
        switch (button.tag) {
            case kButtonTypePlus:
                self.imageButton.transform = CGAffineTransformScale(self.imageButton.transform, 1.2, 1.2);
                break;
                
            case kButtonTypeMinus:
                self.imageButton.transform = CGAffineTransformScale(self.imageButton.transform, 0.8, 0.8);
                break;
        }
    } completion:^(BOOL finished) {
        NSLog(@"%d", finished);
    }];
}

@end

File: CCButtonUse-Swift/CCButtonUse-Swift/ViewController.swift
//
//  ViewController.swift
//		CCButtonUse
//		Chen Chen @ November 2nd, 2016
//

import UIKit

/// 按钮功能枚举数组
enum kButtonType: Int {
    /// 上移按钮
    case top = 10
    /// 下移按钮
    case bottom
    /// 左移按钮
    case left
    /// 右移按钮
    case right
    /// 放大按钮
    case plus
    /// 缩小按钮
    case minus
    /// 左转按钮
    case leftRotate
    /// 右转按钮
    case rightRotate
}

class ViewController: UIViewController {

    /// 按钮使用视图
    var buttonUseView = ButtonUseView(frame: kScreenFrame)
    
    // MARK: - 系统方法
    
    /**
     视图已经加载方法
     */
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        view = buttonUseView
        buttonUseView.topButton.addTarget(self, action: #selector(moveButtonDidClick(button:)), for: .touchUpInside)
        buttonUseView.bottomButton.addTarget(self, action: #selector(moveButtonDidClick(button:)), for: .touchUpInside)
        buttonUseView.leftButton.addTarget(self, action: #selector(moveButtonDidClick(button:)), for: .touchUpInside)
        buttonUseView.rightButton.addTarget(self, action: #selector(moveButtonDidClick(button:)), for: .touchUpInside)
        
        buttonUseView.leftRotateButton.addTarget(self, action: #selector(rotateButtonDidClick(button:)), for: .touchUpInside)
        buttonUseView.rightRotateButton.addTarget(self, action: #selector(rotateButtonDidClick(button:)), for: .touchUpInside)
        
        buttonUseView.plusButton.addTarget(self, action: #selector(scalingButtonDidClick(button:)), for: .touchUpInside)
        buttonUseView.minusButton.addTarget(self, action: #selector(scalingButtonDidClick(button:)), for: .touchUpInside)
    }
    
    // MARK: - 界面方法
    
    /**
     移动按钮点击方法
     */
    func moveButtonDidClick(button: UIButton) {
        
        UIView.beginAnimations(nil, context: nil)
        UIView.setAnimationDuration(1.0)
        switch button.tag {
        case kButtonType.top.rawValue:
            buttonUseView.imageButton.frame.origin.y -= kViewPadding
        case kButtonType.bottom.rawValue:
            buttonUseView.imageButton.frame.origin.y += kViewPadding
        case kButtonType.left.rawValue:
            buttonUseView.imageButton.frame.origin.x -= kViewPadding
        case kButtonType.right.rawValue:
            buttonUseView.imageButton.frame.origin.x += kViewPadding
        default:
            break
        }
        
        UIView.commitAnimations()
    }
    
    /**
     旋转按钮点击方法
     */
    func rotateButtonDidClick(button: UIButton) {
        
        UIView.animate(withDuration: 2.0) {
            switch button.tag {
            case kButtonType.leftRotate.rawValue:
                self.buttonUseView.imageButton.transform = self.buttonUseView.imageButton.transform.rotated(by: (CGFloat)(-M_PI_4))
            case kButtonType.rightRotate.rawValue:
                self.buttonUseView.imageButton.transform = self.buttonUseView.imageButton.transform.rotated(by: CGFloat(M_PI_4))
            default:
                break
            }
        }
    }
    
    /**
     改变按钮点击方法
     */
    func scalingButtonDidClick(button: UIButton) {
        
        UIView.animate(withDuration: 1.0, animations: {
            switch button.tag {
            case kButtonType.plus.rawValue:
                self.buttonUseView.imageButton.transform = self.buttonUseView.imageButton.transform.scaledBy(x: 1.2, y: 1.2)
            case kButtonType.minus.rawValue:
                self.buttonUseView.imageButton.transform = self.buttonUseView.imageButton.transform.scaledBy(x: 0.8, y: 0.8)
            default:
                break
            }
        }) { (result) in
            print(result)
        }
    }
    
}
